php/libraries/SubprojectController.class.inc

```diff
@@ -0,0 +1,252 @@
+<?php
+/**
+ * This file contains the SubprojectController class.
+ *
+ * PHP Version 7
+ *
+ * @category Controller
+ * @package  Subprojects
+ * @author   Rida Abou-Haidar <rida.abou-haidar@mcin.ca>
+ * @license  http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ * @link     https://www.github.com/aces/Loris/
+ */
+
+namespace LORIS;
+
+/**
+ * Base class that contains method to access the subproject of projects
+ *
+ * @category Controller
+ * @package  Subprojects
+ * @author   Rida Abou-Haidar <rida.abou-haidar@mcin.ca>
+ * @license  http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ * @link     https://www.github.com/aces/Loris/
+ */
+class SubprojectController
+{
+    /**
+     * Array holding existing instances
+     *
+     * @var array
+     */
+    private static $_instances = [];
```

Comment:
Using a static variable as a cache can provide economy but it comes with it share of bugs to detect and patch (ex: my comment in the updateSubproject).  I don't think it provides any economy here in most cases. And little economy in few cases. I can't see when multiple subprojects would be instantiated multiple times in the same HTTP request handling. And even if that happened, I am sure it won't be noticeable.

Reply:
So you are proposing to kill the singleton logic ? just reinstanciate when needed ? (I have 0 oppositions, just asking)

Reply:
Technically, the subproject controller is not a singleton. It would need to have  a private constructor and a "getInstance" function to make sur there is only a single static subprojectcontroller in the session. But yes, I think we should stay away from using static variables unless it provides substantial benefits.

Reply:
I meant a factory of subproject singletons. But I see your point and don't mind dropping that